### PR TITLE
33208 Eliminate laminas/laminas-crypt composer dependency

### PR DESCRIPTION
**Pull request by texboy**
### Description (*)

Implement compareStrings function based on the laminas/laminas-crypt implementation
Remove laminas/laminas-crypt dependency from composer.json and update composer.lock

### Related Pull Requests
https://github.com/magento/partners-magento2ee/pull/553

### Fixed Issues (if relevant)

1. Fixes magento/magento2#33208

### Manual testing scenarios (*)

1. Not necessary

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by texboy** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by sivaschenko** [comment]
@magento run all tests

**Comment by sivaschenko** [comment]
Well done @texboy , you made it on the contribution day!

**Comment by sivaschenko** [comment]
The builds have failed due to an issue with the initial branches configuration for the new project. I have fixed it and rerun the tests

**Comment by sivaschenko** [comment]
@texboy would you be able to open a related pull request to the https://github.com/magento/partners-magento2ee repository? (it contains the duplicate of `composer.lock` and `composer.json` that have to be updated as well)

The pull request to EE can be linked to this pull request following the instructions here: https://devdocs.magento.com/contributor-guide/pull-request-tests.html#related-pull-requests

**Comment by texboy** [comment]
@sivaschenko Okay, could you give me access to the repo? Thanks!

**Comment by ihor-sviziev** [comment]
@sivaschenko i think from security perspective it’s better to use laminas crypt here. Usually laminas has much faster release cycle and hotfix for some security issue will be delivered to production much faster

**Comment by Den4ik** [comment]
Sorry for later answer.

> @texboy @sivaschenko i think from security perspective it’s better to use laminas crypt here. Usually laminas has much faster release cycle and hotfix for some security issue will be delivered to production much faster

I'm agree with @ihor-sviziev  that we should continue to use laminal library or switch to native php `hash_equals`

**Comment by Den4ik** [comment]
> Sorry for later answer.
> 
> > @texboy @sivaschenko i think from security perspective it’s better to use laminas crypt here. Usually laminas has much faster release cycle and hotfix for some security issue will be delivered to production much faster
> 
> I'm agree with @ihor-sviziev that we should continue to use laminal library or switch to native php `hash_equals`

Checked changes. I see that code moved from `laminas/crypt` to helper.
I believe that `hash_equals` call is enough because it's part of PHP core.

**Comment by texboy** [comment]
Hello, @Den4ik 

I made the correction you asked.

**Comment by ihor-sviziev** [comment]
Can we cover these changes with some unit test?

**Comment by texboy** [comment]
@ihor-sviziev Luckily, there's already a working unit test for this one :pray: haha
\Magento\Framework\Encryption\Test\Unit\Helper\SecurityTest

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by texboy** [comment]
@sivaschenko 
I created the pull request to the https://github.com/magento/partners-magento2ee  repo and updated the composer dependency. I've linked the pull request.

**Comment by sivaschenko** [comment]
@texboy thank you for creating EE pull request!

@ihor-sviziev @Den4ik laminas-crypt dependency elimination is required for us to be able to provide PHP 8.1 compatible Magento version earlier.

File: lib/internal/Magento/Framework/Encryption/Helper/Security.php [hunk 1]
```diff
@@ -19,12 +15,14 @@ class Security
...
+     * @param string $expected
+     * @param string $actual
      * @return bool
      */
     public static function compareStrings($expected, $actual)
     {
-        return Utils::compareStrings($expected, $actual);
+        $expected     = (string) $expected;
+        $actual       = (string) $actual;
+        return hash_equals($expected, $actual);
```
Comment (sivaschenko) [inline lib/internal/Magento/Framework/Encryption/Helper/Security.php]: @texboy I would simplify this even further and avoid introducing one-time use variables
```suggestion
        return hash_equals((string) $expected, (string) $actual);
```

**Comment by texboy** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]
✔ Approved.

Failing tests look not related to changes from this PR.

**Review by Den4ik — APPROVED** [review]
Agree with @ihor-sviziev 
Looks good for me

**Comment by sivaschenko** [comment]
This is an important improvement for the Platform Health project. Thank you for collaboration @texboy @Den4ik @ihor-sviziev !

**Pull request merged**